File: python_base_04/core/managers/hooks_manager.py
, function_log, game_play_log, log_function_call

class HooksManager:
    def __init__(self):
        # A dictionary to hold hooks and their callbacks with priorities and optional context
        self.hooks = {
            "app_startup": [],  # Predefined default hook
        }
        @log_function_call
    def register_hook(self, hook_name):
        """
        Register a new hook with the given name.
        :param hook_name: str - The name of the hook to register.
        """
        if hook_name in self.hooks:
            raise ValueError(f"Hook '{hook_name}' is already registered.")
        
        self.hooks[hook_name] = []
        @log_function_call
    def register_hook_callback(self, hook_name, callback, priority=10, context=None):
        """
        Register a callback function to a specific hook with a priority and optional context.
        :param hook_name: str - The name of the hook.
        :param callback: function - The callback function to register.
        :param priority: int - The priority of the callback (lower number = higher priority).
        :param context: str - The optional context (e.g., article type) for the callback.
        """
        if hook_name not in self.hooks:
            raise ValueError(f"Hook '{hook_name}' is not registered.")
        
        # Add the callback to the hook
        self.hooks[hook_name].append({
            "priority": priority,
            "callback": callback,
            "context": context
        })
        
        # Sort callbacks by priority
        self.hooks[hook_name].sort(key=lambda x: x["priority"])

        # Detailed logging of the callback registration
        context_info = f" (context: {context})" if context else ""
        callback_name = callback.__name__ if hasattr(callback, "__name__") else str(callback)
        @log_function_call
    def trigger_hook(self, hook_name, data=None, context=None):
        """
        Trigger a specific hook, executing only callbacks matching the context.
        If the hook doesn't exist, it will be automatically registered first.
        :param hook_name: str - The name of the hook to trigger.
        :param data: Any - Optional data to pass to the callbacks.
        :param context: str - The context to filter callbacks (e.g., article type).
        """
        # Auto-register hook if it doesn't exist
        if hook_name not in self.hooks:
            self.register_hook(hook_name)
        
        for entry in self.hooks[hook_name]:
            # Execute only callbacks matching the context or global callbacks (no context)
            if context is None or entry["context"] == context:
                .")
                entry["callback"](data)

    @log_function_call
    def clear_hook(self, hook_name):
        """
        Clear all callbacks registered to a specific hook.
        :param hook_name: str - The name of the hook to clear.
        """
        if hook_name in self.hooks:
            self.hooks[hook_name] = []
            else:
            @log_function_call
    def dispose(self):
        """
        Dispose of all hooks and their callbacks.
        """
        self.hooks.clear()
        

File: dart_bkend_base_01/lib/managers/hooks_manager.dart
/// Represents a hook callback with priority and optional context
class HookCallback {
  final int priority;
  final Function callback;
  final String? context;

  HookCallback(this.callback, this.priority, this.context);
}

/// Hooks Manager - Event-driven callback system matching Python HooksManager
///
/// Enables modules to register callbacks for WebSocket events like room creation,
/// room joining, and room closure. Used for game creation logic and other event-driven features.
class HooksManager {
  final Map<String, List<HookCallback>> _hooks = {
    'app_startup': [], // Predefined default hook
  };

  /// Register a new hook with the given name
  void registerHook(String hookName) {
    if (_hooks.containsKey(hookName)) {
      throw ArgumentError('Hook "$hookName" is already registered.');
    }

    _hooks[hookName] = [];
  }

  /// Register a callback function to a specific hook with priority and optional context
  ///
  /// [hookName] The name of the hook
  /// [callback] The callback function to register
  /// [priority] The priority of the callback (lower number = higher priority)
  /// [context] Optional context for filtering callbacks
  void registerHookCallback(
    String hookName,
    Function callback, {
    int priority = 10,
    String? context,
  }) {
    // Auto-register hook if it doesn't exist (matching Python behavior)
    if (!_hooks.containsKey(hookName)) {
      registerHook(hookName);
    }

    final hookCallback = HookCallback(callback, priority, context);
    _hooks[hookName]!.add(hookCallback);

    // Sort callbacks by priority (lower number = higher priority)
    _hooks[hookName]!.sort((a, b) => a.priority.compareTo(b.priority));
  }

  /// Trigger a specific hook, executing callbacks matching the context
  ///
  /// [hookName] The name of the hook to trigger
  /// [data] Optional data to pass to the callbacks
  /// [context] Optional context to filter callbacks
  void triggerHook(
    String hookName, {
    Map<String, dynamic>? data,
    String? context,
  }) {
    // Auto-register hook if it doesn't exist (matching Python behavior)
    if (!_hooks.containsKey(hookName)) {
      registerHook(hookName);
    }

    final callbacks = _hooks[hookName]!;

    for (final hookCallback in callbacks) {
      // Execute only callbacks matching the context or global callbacks (no context)
      if (context == null || hookCallback.context == context) {
        try {
          // Call the callback with data
          hookCallback.callback(data);
        } catch (e) {
          // Continue executing remaining callbacks (don't break hook chain)
        }
      }
    }
  }

  /// Clear all callbacks registered to a specific hook
  void clearHook(String hookName) {
    if (_hooks.containsKey(hookName)) {
      _hooks[hookName]!.clear();
    }
  }

  /// Get the number of callbacks registered for a hook
  int getHookCallbackCount(String hookName) {
    return _hooks[hookName]?.length ?? 0;
  }

  /// Get all registered hook names
  List<String> getRegisteredHooks() {
    return _hooks.keys.toList();
  }

  /// Check if a hook is registered
  bool isHookRegistered(String hookName) {
    return _hooks.containsKey(hookName);
  }

  /// Dispose of all hooks and their callbacks
  void dispose() {
    _hooks.clear();
  }
}
